Newsletters: add 101 (2020-06-10) (#417)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2020-06-10-newsletter.md

permalink: /en/newsletters/2020/06/10/
name: 2020-06-10-newsletter
slug: 2020-06-10-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter links to a blog post and paper about using
eclipse attacks against Bitcoin software in order to steal from LN
channels and describes the history and impact of a fee overpayment
attack that affects hardware wallets.  Also included are our regular
sections with the summary of a Bitcoin Core PR Review Club meeting,
descriptions of recent software releases, and summaries of recent
changes to popular Bitcoin infrastructure projects.

## Action items

- **Check hardware wallet compatibility:** if your software or your
  processes allow using Trezor devices to spend segwit inputs, check
  whether your system remains compatible with [Trezor's latest firmware
  update][trezor update].  Other hardware wallets may release similar
  updates in the future, contact the manufacturers for more information
  about their plans.  See the *news* item below about the fee
  overpayment attack for more information about the motivation for this
  change.

## News

- **Time dilation attacks against LN:** Gleb Naumenko and Antoine Riard

>   [posted][naumenko td] to the Bitcoin-Dev mailing list a summary of a

Author: How much do you hate this repetition?  I personally found it amusing that they summarized the summary of their paper and phrased this way to convey my amusement, but I'm happy to change it as you suggest if you still want that.

>   [posted][naumenko td] to the Bitcoin-Dev mailing list a summary of a

Reviewer: I must defer to those that have a better sense of humor than I. Keep it.

  public consensus view, the attacker closes their channels with the
  victim in an outdated state that pays the attacker more than the
  latest state would.  The eclipse attack is used to prevent the victim
  from seeing the close transactions until after the dispute period is
  over and the attacker has withdrawn their illicit gains to an address
  entirely under the attacker's control.

    The paper describes an attack that can steal from lightweight
    clients in as little as two hours.  For attacks against LN nodes
    backed by a local Bitcoin full verification node, their shortest
    duration attack requires eclipsing the victim for at least 24 hours.
    The authors note that the attacks probably apply to other
    time-sensitive contract protocols and that the best solutions to the
    problem all involve improving eclipse attack resistance.

- **Fee overpayment attack on multi-input segwit transactions:** The
  transaction fee for a Bitcoin transaction is the difference between
  the amounts of all the UTXOs that transaction spends and the amounts
  of all the UTXOs it creates.  Transactions explicitly state the amounts

>   of the UTXOs they create but the amounts of the UTXOs they spend can

Reviewer: ```suggestion
  of the UTXOs they create, but the amounts of the UTXOs they spend can
```

    Similar to the other attack described in Sander's 2017 email (see
    [Newsletter #97][news97 spk commit]), this attack only affects
    *stateless signers*, such as hardware wallets, which depend on an
    external system to tell them about the UTXOs they control.
    Networked wallets which track the amounts of their received UTXOs
    won't sign for incorrect UTXO amounts and so aren't affected by this
    attack.  Of course, networked wallets are vulnerable to other
    attacks, which is why hardware wallets can enhance user security.

    This past week, Trezor [announced][trezor post] that Saleem Rashid
    rediscovered this vulnerability about three months ago.  In
    response, Trezor has updated its firmware to require copies of
    previous transactions for segwit UTXOs the same way it requires them
    for legacy UTXOs.  This has broken compatibility with several
    wallets that interface with Trezor devices, either directly or
    through [HWI][topic hwi].  In cases where wallets have full copies
    of the previous transactions or are able to obtain them, restoring
    compatibility should just be a matter of updating the wallet code.

>     In other cases where wallets may not store copies of full previous

Author: Changing to "full copies of previous transactions", which sounds better to me and matches the phrasing used in the previous sentence.

      two inputs) that look identical to the user.  Yet the same
      compromised software could show the user two completely different
      transactions (each spending a different input) that look
      identical, resulting in paying the same recipient twice.  Both
      attacks are indistinguishable from the user's perspective but the
      fix for the fee overpayment attack doesn't fix the [spending twice
      attack][maxwell spend twice].

    - **Multisig setups may require multiple compromises:** to perform
      the attack against funds secured by multisig with multiple
      hardware wallets, each signer that is needed to meet the minimal
      threshold (e.g. "2" in a "2-of-3" multisig) would need to be
      tricked into signing the same two transaction variations.  For
      thresholds that include an online wallet which knows the value of
      the UTXO it's signing for (e.g. a policy-based remote signer), the
      attack only works if that online wallet is also compromised.

    A long-term solution to the attack is to change the transaction
    digest so that each signature in a transaction commits to the value

>     of all UTXOs being spent in that transaction, making the signature

Reviewer: It is a dense sentence. Suggest breaking up.
```suggestion
    of all UTXOs being spent in that transaction. This makes the signature
```

Author: With this proposed change, we'd have two sentences in a row that start with "This".  I'll make a slightly more invasive rephrasing to avoid that.

    previous transactions.  Unfortunately, that still doesn't fix the
    spending twice attack, which is an issue with the stateless design
    of most hardware wallets that prevents them from internally tracking
    their own transaction history.

## Bitcoin Core PR Review Club

_In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting._

FIXME:jnewbery or jonatack

{% include functions/details-list.md
  q0="QUESTION"
  a0="ANSWER"
  a0link="/"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 0.20.0][] is the newest major release of Bitcoin's
  oldest project.  This release's most notable user-visible improvements
  include default bech32 addresses for RPC users (the GUI began
  defaulting to bech32 in 0.19), the ability to configure RPC
  permissions for different users and applications (see [Newsletter

>   #77][news77 rpcwhitelist]), some basic support for generating PSBTs in

Author: Agree in principle with the nit, but I still feel that the lack of ability to send PSBTs from the GUI (without using the RPC console) warrants the additional modifier.

  the GUI ([#74][news74 psbtgui] & [#82][news82 psbtgui]), and a
  `sortedmulti` addition to the [output script descriptor][topic
  descriptors] language that can make it easier to watch multisig

>   addresses generated from lexicographical sorted keys, such as those

Reviewer: ```suggestion
  addresses generated from lexicographically sorted keys, such as those
```

  used with ColdCard's native multisig support ([#68][news68
  sortedmulti]).

    Many less visible improvements have also been made to the code in
    order to eliminate bugs, enhance security, and prepare for future
    changes.  One forward-looking change that's attracted some attention
    is the addition of an `asmap` configuration setting that allows

>     using a separately-downloaded database to increase Bitcoin Core's

Reviewer: increase 2x in the sentence
```suggestion
    using a separately-downloaded database to improve Bitcoin Core's
```

    ability to avoid connecting to too many peers all running on a

>     network owned by the same group, increasing it's resistance to

Author: Just checking that you meant so say "its" here (no trailing apostrophe)?

>     network owned by the same group, increasing it's resistance to

Reviewer 2: Indeed; quite the unfortunate typo. s/it's/its/

    unclear how to progress from that.  Gathering and compiling the
    ASN [Autonomous Service Number] data is very nontrivial, and poses
    trust questions."

    For more information about all the changes as well as a list of over
    100 people who contributed to this new version, see the project's
    [release notes][core20 rn].

- [LND 0.10.1-beta][] is a new minor release of this popular LN node
  software.  The release notes announce no major new features but
  do mention several bug fixes, a "new dry run migration mode to test
  out migrations before they're applied permanently, and an enhancement
  to the channel selection/restrictions in the router sub-server."

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Eclair #1440][] Accept multiple channels for some API FIXME:adamjonas

- [Eclair #1141][] adds support for `option_static_remotekey` channels.

>   In the case of data loss on your end, this feature allows your channel

Reviewer 2: Note: There are two places where the 3rd-person user is switched to 2nd-person "you" -- this paragraph, and the "Check hardware wallet compatibility" paragraph in Action Items.

Author: I'm changing this paragraph to third person, but the Action Items section is definitely meant to refer directly to the reader.

  counterparty to pay an untweaked key agreed upon during the initial
  channel open, which will close the channel and allow your wallet to
  spend your funds. See [Newsletter #67][news67 static_remotekey] for
  more details on this feature.

- [LND #4251][] REST saga 2/3: Add REST endpoints to all RPCs FIXME:dongcarl

> - [BIPs #920][] updates the [BIP341][] specification of [taproot][topic

Reviewer 2: s/BIPs/BIP/

Author: BIPs is "correct" here, at least per the convention I've been using in this section for more than a year (see `git grep '\[BIPs #'`).  I'm trying to convey that this is `BIPs [repository PR] #920` and not "BIP number 920" (which we would write as BIP920).

Reviewer 2: Oh right, the repo :hammer:

    Hardware wallets don't maintain a UTXO set, so the only trustless
    way for them to determine the amount of fee paid by a transaction
    with legacy UTXOs is to get a copy of each legacy UTXO's previous
    transaction, hash the previous transaction to ensure its txid
    matches the UTXO reference, and use the now-verified amount of the
    UTXO to perform the fee calculation.  Since legacy transactions can
    be almost as large as one megabyte and it's possible for a spending
    transaction to reference thousands of previous transactions, this
    process can require the processing of gigabytes of data by
    resource-constrained hardware wallets.

    One of the several improvements made in [BIP143][] segwit v0 was an
    attempt to eliminate this burden by having [signatures commit to the
    amount][bip143 motivation] of the UTXO they were spending.  This
    meant that any signatures that committed to an incorrect amount
    would be invalid, which developers of both [Bitcoin Core][towns
    benefits] and [hardware wallets][palatinus inpatient] believed would
    allow signers to safely accept amounts from untrusted programs.

>     Unfortunately, signing a single UTXO's amount turned out to be

Reviewer 3: I'd emphasize that the 'attacker' is not able to directly steal any funds from the victim, simply make him burn bitcoin in fees.

    insufficient.  In 2017, Greg Sanders [described][sanders attack] how
    an attacker could send two versions of the same transaction with two
    segwit UTXOs to a hardware wallet.  In the first case, the attacker
    would understate the amount of one of the UTXOs being spent; the

>     second case, they'd understate the amount of the second UTXO.

Reviewer 3: either "one of the UTXOs being spent ... the other UTXO being spent" or "the first UTXO being spent ... the second UTXO being spent"

Author: It appears I pushed an update after you began reviewing this; I think that addressed your comment.  Could you please re-review?

Reviewer 3: yes, this is now addressed.

    Because both transactions understated the amount of one of the
    UTXOs, the hardware wallet would under-calculate the fee by that
    same understated amount.  A user would authorize signing the first
    transaction based on its understated fee.  The attacker would then

>     claim there was some minor problem and the user needs to sign the

Reviewer 3: nit: would claim _that_ there was .... and _that_ the user?

    same transaction again, but would instead get the user to sign the

>     second transaction (which would look identical to the user).

Reviewer 3: "would look identical to the user" is confusing since the transactions are not identical. How about "which would appear to be the same to the user since they pay the same outputs and same fee"

    Although both transactions produced are individually invalid because

>     they each contain one signature that commits to a wrong amount, each

Reviewer 3: I'd remove the part about the invalid signature. The attacker doesn't need to provide a signature at all.

Author: I don't understand this comment.

Reviewer 3: Oh, I misunderstood. I didn't understand that both UTXOs are owned by the victim. Perhaps that needs to be explicitly stated for slow people like me?

    transaction also contains one signature that's valid.  The attacker
    uses those two valid signatures to synthesize a valid transaction
    that overpays its fees.

    ![Fee overpayment attack illustration](/img/posts/2020-06-fee-overpayment-attack.dot.png)

    Similar to the other attack described in Sander's 2017 email (see

>     [Newsletter #97][news97 spk commit]), this attack only affects

Reviewer 4: OOC what was newly revealed this week vs. Sander's 2017 email? Why wasn't this discussed more and addressed 3 years ago?

Author: I don't know what "OOC" means, but as far as I know nothing new has been revealed this week.  I think it was discussed a bit more back then (e.g. on IRC) and this summary describes the attempts to resolve it at the protocol layer (e.g. Lau's SIGHASH2 proposal in 2018 and bip-taproot in 2019, both of which we summarized in Optech newsletters).

It appears from Trezor's description that they just learned about this attack when Rashid reported it and I'm guessing that that's also when other hardware wallets learned about it (or through Trezor's own disclosure, as it appears Rashid didn't contact Coinkite (ColdCard) or some other hardware wallet vendors).  I've been aware of this attack since shortly after Sander's published his description (I think I first heard about it on IRC) and it never occurred to me that it wasn't known to vendors.

I have a pet theory that Trezor didn't see Sanders's post because it occurred in a thread about PSBT and Trezor has historically been [opposed](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013041.html) to hardware wallet interface standardization, Ledger has never been really involved in Bitcoin protocol development, and CoinKite hadn't started work on ColdCard at that point.

As to why nobody else worked on it, comments in #bitcoin-core-dev IRC indicate that nobody who knew about the attack for the past three years thought it was particularly urgent.  I'll update to sprinkle a few more links through the text.  My own personal take on it is that it isn't that severe of an issue but that probably all hardware wallet manufacturers will end up doing the same thing Trezor does just to prevent FUD from other HW manufacturers and to cover their asses against people claiming to lose money due to this vulnerability (whether legitimate or illegitimate---if you're a miner or know a trusted miner, it's easy to fake being a victim of this attack).

Reviewer 2: @harding Well said. This is my understanding as well, from the various IRC discussions of the past week.

Reviewer 2: At the last wallet meeting Friday, people weren't immediately convinced it was worth patching: http://www.erisian.com.au/bitcoin-core-dev/log-2020-06-05.html#l-346

      attack][maxwell spend twice].

    - **Multisig setups may require multiple compromises:** to perform
      the attack against funds secured by multisig with multiple
      hardware wallets, each signer that is needed to meet the minimal
      threshold (e.g., "2" in a "2-of-3" multisig) would need to be
      tricked into signing the same two transaction variations.  For
      thresholds that include an online wallet which knows the value of
      the UTXO it's signing for (e.g., a policy-based remote signer), the
      attack only works if that online wallet is also compromised.

    A long-term solution to the attack is to change the transaction
    digest so that each signature in a transaction commits to the value
    of all UTXOs being spent in that transaction, making the signature
    invalid if an attacker lies about the amount of any UTXO.  This was
    [proposed][lau sighash2] by Johnson Lau in 2018 (see [Newsletter
    #11][news11 sighash]) and has been included in the [BIP341][]
    specification of [taproot][topic taproot] since its earliest public
    drafts (see [Newsletter #46][news46 digest changes]).  If taproot is

>     adopted, then it should become perfectly safe for stateless signers

Reviewer 3: Drop "perfectly". Claiming perfection is always dangerous!

    like hardware wallets to sign for taproot UTXOs without referencing
    previous transactions.  Unfortunately, that still doesn't fix the
    spending twice attack, which is an issue with the stateless design
    of most hardware wallets that prevents them from internally tracking
    their own transaction history.

> 

Reviewer 4: Should we also call out the software that is now broken such as BTCPay server and Wasabi if a user upgrades to the latest Trezor firmware? That is another drawback/tradeoff.

Author: We do this, we just don't explictly name the software that's broken.

> This has broken compatibility with several wallets that interface with Trezor devices, either directly or through HWI.

    attack][] (see [Newsletter #83][news83 asmap]).  However, one of the
    authors of the feature
    [notes][wuille asmap], "it's highly experimental for now, and
    unclear how to progress from that.  Gathering and compiling the
    ASN [Autonomous Service Number] data is very nontrivial, and poses
    trust questions."

    For more information about all the changes as well as a list of over
    100 people who contributed to this new version, see the project's
    [release notes][core20 rn].

- [LND 0.10.1-beta][] is a new minor release of this popular LN node
  software.  The release notes announce no major new features but
  do mention several bug fixes, a "new dry run migration mode to test
  out migrations before they're applied permanently, and an enhancement
  to the channel selection/restrictions in the router sub-server."

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

> - [Eclair #1440][] allows for updating multiple channels via the API. Proposed

Reviewer 3: "allows for updating" sounds a little odd to me (the primary definition of "allow for" is to plan for something that might or might not happen (https://www.merriam-webster.com/dictionary/allow%20for). How about "adds support for updating" or "allows multiple channels to be updated"

  [posted][naumenko td] to the Bitcoin-Dev mailing list a summary of a
  [blog post][td blog] they wrote that itself summarizes a [paper][td
  paper] they authored about using [eclipse attacks][topic eclipse attacks]
  to steal money from LN channels.  This extends the analysis by Riard
  described in [Newsletter #77][news77 eclipse].  In short, an attacker
  gains control over all of an LN node's connections to the Bitcoin P2P
  network and delays the relay of new block announcements to the victim.
  After the victim's view of the block chain is far enough behind the
  public consensus view, the attacker closes their channels with the
  victim in an outdated state that pays the attacker more than the
  latest state would.  The eclipse attack is used to prevent the victim
  from seeing the close transactions until after the dispute period is
  over and the attacker has withdrawn their illicit gains to an address
  entirely under the attacker's control.

    The paper describes an attack that can steal from lightweight
    clients in as little as two hours.  For attacks against LN nodes
    backed by a local Bitcoin full verification node, their shortest

>     duration attack requires eclipsing the victim for at least 24 hours.

Reviewer 5: In fact that's few hours, depending on implementation time-locks (see Table 1 : Default timelock (in blocks), timeout policy). 

You can also note, due to the lack of hardened eclipse protection in lightweight client and low-number of deployed lightweight servers, this attack is practical with high-chances of success _today_, assuming you have a channel with the victim.

    The authors note that the attacks probably apply to other
    time-sensitive contract protocols and that the best solutions to the

>     problem all involve improving eclipse attack resistance.

Reviewer 5: The AltNet prroposal (https://github.com/bitcoin/bitcoin/pull/18988) has been quite motivated by this research. The current base layer p2p assumptions are limited with regards to LN security model.

Reviewer 5: Blog post invite future research on inter-layer mapping or native-LN attacks. For more context, see IRC logs of LN dev meeting of today, which is about relevance of those topics.

name: 2020-06-10-newsletter
slug: 2020-06-10-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter links to a blog post and paper about using
eclipse attacks against Bitcoin software in order to steal from LN
channels and describes the history and impact of a fee overpayment
attack that affects hardware wallets.  Also included are our regular
sections with the summary of a Bitcoin Core PR Review Club meeting,
descriptions of recent software releases, and summaries of recent
changes to popular Bitcoin infrastructure projects.

## Action items

- **Check hardware wallet compatibility:** if your software or your
  processes allow using hardware wallets to spend segwit inputs, check
  whether your system remains compatible with your wallet's latest
  firmware update.  For example, [Trezor's latest firmware][trezor
  update] requires software interfacing with it to be upgraded to
  handle segwit inputs; [Ledger's latest firmware][ledger update] will
  print a warning when processing segwit inputs received from
  non-updated software.  Other hardware wallets may release similar
  updates in the future; contact the manufacturers for more information

>   about their plans.  See the *news* item below about the fee

Reviewer 2: perhaps s/news/News/ and s/item/section/

  from seeing the close transactions until after the dispute period is
  over and the attacker has withdrawn their illicit gains to an address
  entirely under the attacker's control.

    The paper describes an attack that can steal from lightweight
    clients in as little as two hours.  For attacks against LN nodes
    backed by a local Bitcoin full verification node, their shortest
    duration attack requires eclipsing the victim for at least 24 hours.
    The authors note that the attacks probably apply to other
    time-sensitive contract protocols and that the best solutions to the
    problem all involve improving eclipse attack resistance.

- **Fee overpayment attack on multi-input segwit transactions:** The
  fee paid by a Bitcoin transaction is the difference between
  the amounts of the UTXOs the transaction spends and the amounts
  of the UTXOs it creates.  Transactions explicitly state the amounts
  of the UTXOs they create, but the amounts of the UTXOs they spend can
  only be found by looking at the previous transactions that created
  those UTXOs.  The spending transaction simply commits to the txid and

>   location of the UTXOs it wants to spend, requiring other software that

Author: Changing s/location/position/.  "Output index (vout)" is *exactly* the phrasing I'd use if I wanted to be precise, but in this case I just want to explain that the hardware wallet needs the full previous transaction (sans witness) in order to calculate the txid.

    wallets that interface with Trezor devices, either directly or
    through [HWI][topic hwi].  In cases where wallets have full copies
    of the previous transactions or are able to obtain them, restoring
    compatibility should just be a matter of updating the wallet code.
    In other cases where wallets may not store full copies of previous
    transactions that paid the wallet, this may require redesigning the
    wallet to store that data plus require rescanning the block chain
    for past wallet transactions.  Software that creates or updates
    [Partially Signed Bitcoin Transactions][topic psbt] (PSBTs) will
    need to be [upgraded][chow full input tx] to include the full copies
    of previous transactions into the PSBTs.  For cases where PSBTs are
    stored in size-constrained
    media (e.g., QR codes, see [Newsletter #96][news96 qr]), the
    potentially significant increase in data size may require abandoning
    the protocol or switching to higher-capacity media.

    Ledger has made a similar change, although their [support
    article][ledger update] says they only display a warning when
    spending segwit UTXOs without full copies of their previous

>     transactions.  Although requiring previous transactions does

Author: I think you missed a golden opportunity to start your comment by saying: "Although all those althoughs..."  :-)

    maximize user safety against this attack, there are also arguments

>     for allowing hardware wallets to optionally continue to use segwit's

Reviewer 2: perhaps s/to use/using/ to align with "rather than breaking"

Reviewer 2: Bitcoin Core PR opened yesterday on this by Andrew Chow: https://github.com/bitcoin/bitcoin/pull/19215

As well as in HWI: https://github.com/bitcoin-core/HWI/pull/340

projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 0.20.0][] is the newest major release from the original
  Bitcoin project.  This release's most notable user-visible improvements
  include default bech32 addresses for RPC users (the GUI began
  defaulting to bech32 in 0.19), the ability to configure RPC
  permissions for different users and applications (see [Newsletter
  #77][news77 rpcwhitelist]), some basic support for generating PSBTs in
  the GUI ([#74][news74 psbtgui] & [#82][news82 psbtgui]), and a
  `sortedmulti` addition to the [output script descriptor][topic
  descriptors] language that can make it easier to watch multisig
  addresses generated from lexicographically sorted keys, such as those
  used with ColdCard's native multisig support ([#68][news68
  sortedmulti]).

    Many less visible improvements have also been made to the code in
    order to eliminate bugs, enhance security, and prepare for future
    changes.  One forward-looking change that's attracted some attention
    is the addition of an `asmap` configuration setting that allows

>     using a separately-downloaded database to improve Bitcoin Core's

Reviewer 2: nit: can omit hyphen

    ability to avoid connecting to too many peers all running on a
    network owned by the same group, increasing its resistance to
    [eclipse attacks][topic eclipse attacks] such as the [Erebus

>     attack][] (see [Newsletter #83][news83 asmap]).  However, one of the

Reviewer 2: Likely depends on who one talks to, but I'm not sure the terminology and description is ideal:

- The asmap feature and discussion specifically targeted Erebus: see https://github.com/bitcoin/bitcoin/pull/16702 and https://github.com/bitcoin/bitcoin/issues/16599.

- The researchers may be biased but do describe Erebus as a separate, third partitioning attack (see https://erebus-attack.comp.nus.edu.sg, section "How is the Erebus attack different from previous partitioning attacks?"): *There are two previous partitioning attacks, i.e., the Bitcoin hijacking (by Apostolaski et al., IEEE SP'17) and the Eclipse attack (by Heilman et al., USENIX Sec'15)*, followed by a table that details the distinguishing characteristics.

Reviewer 2: Perhaps the question is whether "eclipse attacks" has become a generic term for partitioning attacks, separate from the Eclipse attack (2015).

Author: @jonatack Eclipse Attack is a term that predates Bitcoin.  Here's a paper from 2006: https://homepage.divms.uiowa.edu/~ghosh/eclipse.pdf .  I don't claim to be an expert, but I think eclipse attack is the generic name for isolating a node from any honest peer so that you can feed it misinformation.  Heilman's paper describe several *methods* for eclipsing a Bitcoin node, and the Erebus attack describes another method.  In short, I think the Erebus attack is a type of eclipse attack (and I previously updated the text to say that).

Reviewer 2: > @jonatack Eclipse Attack is a term that predates Bitcoin. Here's a paper from 2006

TIL - thanks! Will read.

[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1 repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

- [Eclair #1440][] allows for updating multiple channels via the API. Proposed
  initially to help node operators adjust to the dynamic on-chain fee market
  by batch updating their channels' relay fees, this addition enables batch
  closing and force-closing as well.

- [Eclair #1141][] adds support for `option_static_remotekey` channels.
  In the case of data loss by an LN node, this feature allows the channel
  counterparty to pay the current balance to an untweaked key agreed upon during the initial
  channel open, closing the channel and allowing the node's wallet to
  spend its funds normally. See [Newsletter #67][news67 static_remotekey] for
  more details on this feature.

- [LND #4251][] brings its REST interface's list of supported non-streaming
  messages to parity with that of its gRPC interface. Previously, only a subset
  of these messages (specifically those of the `Lightning` subserver) were

>   available via the REST interface. This is PR will be followed by [LND

Reviewer 2: s/is PR/PR/

    compromised software creates two variations of the requested
    transaction that each spend the same two segwit UTXOs controlled by
    the user's attached hardware wallet.  In the first transaction, the
    compromised software understates the amount of one of the UTXOs
    being spent.  This will cause the hardware wallet to under-calculate
    the transaction fee by that same understated amount.  In the second
    transaction, the compromised software understates the amount of the
    second UTXO.  That will cause the hardware wallet to also
    under-calculate the second transaction's fee.

    The compromised software sends the first transaction to the hardware
    wallet.  The user checks the amount, receiver address, and calculated
    transaction fee.  The user authorizes signing and the signed
    transaction is returned to the compromised software.  This first
    transaction is invalid under BIP143 because one of its signatures
    commits to an incorrect UTXO amount.  The compromised software then
    claims that there was some minor problem that prevented the
    transaction from being broadcast and tells the user they need to

>     re-sign the transaction, but the wallet doesn't re-send the first

Reviewer 3: s/wallet/software/. You've distinguished between 'hardware wallet' and 'compromised software' up to this point. Using the word wallet here is confusing.

    won't sign for incorrect UTXO amounts and so aren't affected by this
    attack.  Of course, networked wallets are vulnerable to other
    attacks, which is why hardware wallets can enhance user security.

    This past week, Trezor [announced][trezor post] that Saleem Rashid
    rediscovered this vulnerability about three months ago.  In
    response, Trezor has updated its firmware to require copies of
    previous transactions for segwit UTXOs the same way it requires them
    for legacy UTXOs.  This has broken compatibility with several
    wallets that interface with Trezor devices, either directly or
    through [HWI][topic hwi] (see [HWI #340][]).  In cases where wallets have full copies
    of the previous transactions or are able to obtain them, restoring
    compatibility should just be a matter of updating the wallet code.
    In other cases where wallets may not store full copies of previous
    transactions that paid the wallet, this may require redesigning the
    wallet to store that data plus require rescanning the block chain
    for past wallet transactions.  Software that creates or updates
    [Partially Signed Bitcoin Transactions][topic psbt] (PSBTs) will

>     need to be [upgraded][Bitcoin Core #12915] to include the full copies

Author: Transposition :facepalm: Thanks!

      and blocks, desirable service flags, and long-lasting connections. Of the
      remaining peers, choose one to evict from the netgroup with the most
      connections."
  a3link="https://github.com/bitcoin-core-review-club/bitcoin/blob/pr16756/src/net.cpp#L851"
%}

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [Bitcoin Core 0.20.0][] is the newest major release from the original
  Bitcoin project.  This release's most notable user-visible improvements
  include default bech32 addresses for RPC users (the GUI began
  defaulting to bech32 in 0.19), the ability to configure RPC
  permissions for different users and applications (see [Newsletter
  #77][news77 rpcwhitelist]), some basic support for generating PSBTs in
  the GUI ([#74][news74 psbtgui] & [#82][news82 psbtgui]), and a
  `sortedmulti` addition to the [output script descriptor][topic
  descriptors] language that can make it easier to watch multisig
  addresses generated from lexicographically sorted keys, such as those

>   used with ColdCard's native multisig support ([#68][news68

Reviewer 6: s/ColdCard/Coldcard

  wallets to safely participate in coinjoins and other collaboratively
  generated transactions.  For details, see the description in
  [Newsletter #97][news97 spk commit].

## Special thanks

We thank Pieter Wuille for assistance researching the history of the fee
overpayment attack.  We thank Antoine Riard for additional information
about the time dilation attack paper.  Any errors or omissions are the fault of the
newsletter author.

{% include references.md %}
{% include linkers/issues.md issues="1440,4251,920,1141,4141,16756,18988,12915" %}
[bitcoin core 0.20.0]: https://bitcoincore.org/bin/bitcoin-core-0.20.0
[lnd 0.10.1-beta]: https://github.com/lightningnetwork/lnd/releases/tag/v0.10.1-beta
[news97 spk commit]: /en/newsletters/2020/05/13/#request-for-an-additional-taproot-signature-commitment
[news77 eclipse]: /en/newsletters/2019/12/18/#discussion-of-eclipse-attacks-on-ln-nodes
[hw security]: https://en.bitcoin.it/wiki/Hardware_wallet#Security_risks
[towns benefits]: https://bitcoincore.org/en/2016/01/26/segwit-benefits/#signing-of-input-values
[palatinus inpatient]: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-October/013248.html

> [news11 sighash]: https://bitcoinops.org/en/newsletters/2018/09/04/#proposed-sighash-updates

Reviewer 6: s/
https://bitcoinops.org/en/newsletters/2018/09/04/#proposed-sighash-updates
/en/newsletters/2018/09/04/#proposed-sighash-updates
